File: solutions/golang/200/solution.go
var directions = [][]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}
var seen [][]bool
var M int
var N int

func numIslands(grid [][]byte) int {
	M = len(grid)
	N = len(grid[0])
	seen = make([][]bool, M)
	for i := 0; i < M; i++ {
		seen[i] = make([]bool, N)
	}

	ans := 0
	for i := 0; i < M; i++ {
		for j := 0; j < N; j++ {
			if grid[i][j] == '1' && seen[i][j] == false {
				seen[i][j] = true
				ans++
				dfs(i, j, grid)
			}
		}
	}

	return ans
}

func dfs(row int, col int, grid [][]byte) {
	for _, dir := range directions {
		newRow := row + dir[0]
		newCol := col + dir[1]
		if valid(newRow, newCol, grid) && seen[newRow][newCol] == false {
			seen[newRow][newCol] = true
			dfs(newRow, newCol, grid)
		}
	}
}

func valid(row int, col int, grid [][]byte) bool {
	return row >= 0 && row < M && col >= 0 && col < N && grid[row][col] == '1'
}

File: solutions/cpp/200/solution.cpp
class Solution {
public:
    vector<vector<int>> DIRECTIONS = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    int M;
    int N;
    vector<vector<bool>> seen;
    int numIslands(vector<vector<char>>& grid) {
        this->M = grid.size();
        this->N = grid[0].size();
        seen = vector(M, vector(N, false));

        int ans = 0;
        for (int row = 0; row < M; row++) {
            for (int col = 0; col < N; col++) {
                if (seen[row][col] == true || grid[row][col] == '0') {
                    continue;
                }

                seen[row][col] = true;
                dfs(row, col, grid);
                ans++;
            }
        }
        return ans;
    }

    void dfs(int row, int col, vector<vector<char>>& grid) {
        for (vector<int> &dir : DIRECTIONS) {
            int newRow = row + dir[0];
            int newCol = col + dir[1];
            if (valid(newRow, newCol, grid) && !seen[newRow][newCol]) {
                seen[newRow][newCol] = true;
                dfs(newRow, newCol, grid);
            }
        }
    }

    bool valid(int row, int col, vector<vector<char>>& grid) {
        return row >= 0 && row < M && col >= 0 && col < N && grid[row][col] == '1';
    }
};